test/distributed/_tensor/test_dtensor_compile.py

```diff
@@ -182,20 +186,40 @@ def world_size(self):
         return 4
 
     @with_comms
-    def test_tp_compile_fullgraph(self):
+    @parametrize("is_seq_parallel", [True, False])
+    def test_tp_compile_fullgraph(self, is_seq_parallel):
         mesh = DeviceMesh(self.device_type, torch.arange(self.world_size))
 
         model = SimpleModel(self.device_type)
+        module_prepare_input = (
+            PrepareModuleInput()
+            if is_seq_parallel
+            else PrepareModuleInput(input_layouts=Replicate())
+        )
+        no_input_prepare_colwise_style = ColwiseParallel(input_layouts=None)
+        colwise_style = (
+            ColwiseParallel(input_layouts=Shard(0))
+            if is_seq_parallel
+            else ColwiseParallel()
+        )
+        rowwise_style = (
+            RowwiseParallel(output_layouts=Shard(0))
+            if is_seq_parallel
+            else RowwiseParallel()
+        )
         model = parallelize_module(
             model,
             mesh,
             parallelize_plan={
-                "mlp_0.net1": ColwiseParallel(),
-                "mlp_0.net2": RowwiseParallel(),
-                "mlp_1.net1": ColwiseParallel(),
-                "mlp_1.net2": RowwiseParallel(),
+                "mlp_0": module_prepare_input,
```

Comment:
i'm not sure what this does- does it mean that input is already coming as local tensors (may be sharded or replicated) and this wraps them in DTensors?  Or does this actually 'shard' the inputs (Assumes whole inputs first)

Reply:
> input is already coming as local tensors (may be sharded or replicated) and this wraps them in DTensors

Yes. So what I see in xlformer is that it has three/two col-wise linear. Instead of registering hook in each nn.Linear and end up calling all-gather multiple times, we will use `module_prepare_input` to only register once in the parent module.

Reply:
@wconstab it's coming in as local tensors then in this prepareInput we mark it as DTensor and do a allgather (redistribute)

@fduwjj another follow up we should probably do is to remove the default `layouts` for PrepareModuleInput/Output, and requires user to set it more explicitly so that user know what things is being done under the hood